server/src/prefect_server/services/zombie_killer/zombie_killer.py

```diff
@@ -0,0 +1,140 @@
+
+    Zombies are tasks that should be checkpointing with the Orchestration
+    layer, but due to some issue (typically but not exclusively network issues), they
+    are unable to. They're called zombies because they could still be running just fine,
+    but since they're unable to communicate with the Orchestration layer, we have
+    to assume they're dead in order to keep flows running smoothly.
+
+    Zombies are considered as:
+        - A running task run
+        - the flow associated with the task run does has heartbeats enabled
+        - the task run hasn't checked in in 2 minutes.
+
+    The Zombie Killer loads up to 100 task runs, matching these conditions, prioritizing tasks
+    by start time. These task runs are marked are then moved from their current state
+    to a Failed state with a message noting that it was killed by the ZombieKiller.
+    """
+
+    zombie_killer_config_key = "services.zombie_killer.zombie_killer_loop_seconds"
+    # Scheduled to run relatively frequently. Set to longer if this is putting too
+    # much stress on your database.
+    loop_seconds_default = 120
+
+    async def kill_zombies(self, n_zombies: int = 100) -> int:
+        """
+        Main function of the Zombie Killer, responsible for locating the 
+        correct task runs to kill. Task runs are eligible to be killed if they
+        meet all criteria as a "Zombie".
+
+        Args:
+            - n_zombies (int): Maximum number of task runs to kill
+
+        Returns:
+            - int: Number of task runs actually killed
+        """
+        kill_threshold = pendulum.now("UTC") - pendulum.duration(minutes=2)
+
+        zombie_task_runs = await models.TaskRun.where(
+            {
+                # Task run is currently running
+                "state": {"_eq": "Running"},
+                "_and": [
+                    {
+                        "_or": [
+                            # hasn't had a heartbeat since the threshold
+                            {"heartbeat": {"_lte": kill_threshold.isoformat()}},
+                            # Hasn't had a heartbeat since task run creation, in which
+                            # case we can safely default back to `updated`. If we use
+                            # created, we can possibly end up in a loop of a task surviving
+                            # and never heartbeating, and every time it would get rescheduled,
+                            # the zombie killer would kill it immediately.
+                            {
+                                "_and": [
+                                    {"heartbeat": {"_is_null": True}},
+                                    {"updated": {"_lte": kill_threshold.isoformat()}},
+                                ]
+                            },
+                        ],
+                    },
+                    {  # For backwards compatability, we check whether the
```

Comment:
This is 💯 but elsewhere in Server we've made the decision not to support versions of Prefect < 0.10.0 (the release that introduced Server) because the burden is too high. I think there's no harm in keeping it in, but just calling it out